picasso/src/main/java/com/squareup/picasso/MarkableInputStream.java

```diff
@@ -127,6 +131,9 @@ private void skip(long current, long target) throws IOException {
   }
 
   @Override public int read(byte[] buffer) throws IOException {
+    if (offset + buffer.length >= limit) {
+      setLimit(2 * (offset + buffer.length));
```

Comment:
So this behavior isn't quite right. We need the buffering of data to only happen on when a marker is set (for reading the image metadata) but to NOT happen when there is no marker (during the actual full image decode). As it stands now this will cause a ton of allocation and GC churn in the latter case when we don't actually care about buffering.

My approach was going to be killing this class for another `InputStream` decorator that used an on-demand list of 2K `byte[]` chunks for infinite buffering when a mark was set. In the ideal case only a single 2K chunk would be allocated. In the pathological degradation case it would just allocate additional chunks as needed until reset.